stl/inc/ranges

```diff
@@ -435,6 +438,197 @@ namespace ranges {
+                } else {
+                    // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        // clang-format off
+        _Defaultabox& operator=(_Defaultabox&&) noexcept
+            requires is_trivially_move_assignable_v<_Ty> = default;
+        // clang-format on
+
+        _CONSTEXPR20_DYNALLOC _Defaultabox& operator=(_Defaultabox&& _That) noexcept(
+            is_nothrow_move_constructible_v<_Ty>&& is_nothrow_move_assignable_v<_Ty>) /* strengthened */ {
+            if (_Engaged) {
+                if (_That._Engaged) {
+                    _Val = _STD move(_That._Val);
+                } else {
+                    _Val.~_Ty();
+                    _Engaged = false;
+                }
+            } else {
+                if (_That._Engaged) {
+                    _Construct_in_place(_Val, _STD move(_That._Val));
+                    _Engaged = true;
+                } else {
+                    // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        constexpr _Defaultabox& operator=(_Ty&& _That) noexcept(
+            is_nothrow_move_constructible_v<_Ty>&& is_nothrow_move_assignable_v<_Ty>) {
+            if (_Engaged) {
+                _Val = _STD move(_That);
+            } else {
+                _Construct_in_place(_Val, _STD move(_That));
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        constexpr _Defaultabox& operator=(const _Ty& _That) noexcept(
+            is_nothrow_copy_constructible_v<_Ty>&& is_nothrow_copy_assignable_v<_Ty>) requires copyable<_Ty> {
+            if (_Engaged) {
+                _Val = _That;
+            } else {
+                _Construct_in_place(_Val, _That);
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return _Engaged;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        constexpr void reset() noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+        }
+
+        constexpr bool operator==(const _Defaultabox& _That) const
+            noexcept(noexcept(_Val == _That._Val)) /* requires equality_comparable<_Ty> */ {
+            if (_Engaged) {
+                if (_That._Engaged) {
+                    return _Val == _That._Val;
+                } else {
+                    return false;
+                }
+            } else {
+                if (_That._Engaged) {
+                    return false;
+                } else {
+                    return true;
+                }

```

Comment:
I figured this could be simplified - In fact the entire body is `return _Engaged == _That._Engaged && (!_Engaged || _Val == _That._Val);` - but kept the structure for consistency. Looking at it now, I think the consistency benefit is outweighed by the cost in LOC. I'll simplify.